lxd/main_init.go

```diff
@@ -273,44 +281,256 @@ they otherwise would.
+		reverter, err := changer()
 		if err != nil {
+			cmd.revert(reverters)
 			return err
 		}
+		// Save the revert function for later.
+		reverters = append(reverters, reverter)
+	}
 
-		if trustPassword != "" {
-			err = cmd.setServerConfig(c, "core.trust_password", trustPassword)
-			if err != nil {
-				return err
-			}
+	return nil
+}
+
+// Try to revert the state to what it was before running the "lxd init" command.
+func (cmd *CmdInit) revert(reverters []reverter) {
+	for _, reverter := range reverters {
+		err := reverter()
+		if err != nil {
+			logger.Warnf("Reverting to pre-init state failed: %s", err)
+			break
 		}
 	}
+}
 
-	cmd.Context.Output("LXD has been successfully configured.\n")
-	return nil
+// Apply the server-level configuration in the given map.
+func (cmd *CmdInit) initConfig(client lxd.ContainerServer, config map[string]interface{}) (reverter, error) {
+	server, etag, err := client.GetServer()
+	if err != nil {
+		return nil, err
+	}
+
+	// Build a function that can be used to revert the config to
+	// its original values.
+	reverter := func() error {
+		return client.UpdateServer(server.Writable(), "")
+	}
+
+	// The underlying code expects all values to be string, even if when
+	// using preseed the yaml.v2 package unmarshals them as integers.
+	for key, value := range config {
+		if number, ok := value.(int); ok {
+			value = strconv.Itoa(number)
+		}
+		config[key] = value
+	}
+
+	err = client.UpdateServer(api.ServerPut{Config: config}, etag)
+	if err != nil {
+		return nil, err
+	}
+
+	// Updating the server was sucessful, so return the reverter function
+	// in case it's needed later.
+	return reverter, nil
+}
+
+// Create or update a single pool, and return a revert function in case of success.
+func (cmd *CmdInit) initPool(client lxd.ContainerServer, pool api.StoragePoolsPost) (reverter, error) {
+	var reverter func() error
+	currentPool, _, err := client.GetStoragePool(pool.Name)
+	if err == nil {
+		reverter, err = cmd.initPoolUpdate(client, pool, currentPool.Writable())
+	} else {
+		reverter, err = cmd.initPoolCreate(client, pool)
+	}
+	if err != nil {
+		return nil, err
+	}
+	return reverter, nil
+}
+
+// Create a single new pool, and return a revert function to delete it.
+func (cmd *CmdInit) initPoolCreate(client lxd.ContainerServer, pool api.StoragePoolsPost) (reverter, error) {
+	reverter := func() error {
+		return client.DeleteStoragePool(pool.Name)
+	}
+	err := client.CreateStoragePool(pool)
+	return reverter, err
+}
+
+// Update a single pool, and return a function that can be used to
+// revert it to its original state.
+func (cmd *CmdInit) initPoolUpdate(client lxd.ContainerServer, pool api.StoragePoolsPost, currentPool api.StoragePoolPut) (reverter, error) {
+	reverter := func() error {
+		return client.UpdateStoragePool(pool.Name, currentPool, "")
+	}
+	err := client.UpdateStoragePool(pool.Name, api.StoragePoolPut{
+		Config: pool.Config,
+	}, "")
+	return reverter, err
+}
```

Comment:
The storage pool API doesn't exist in LXD 2.0, so I don't see how initPoolCreate and initPoolUpdate can possibly work.

Reply:
Good catch, this is artifact of the cherry pick, but effectively idle code that can't be reached. Removed.